### PR TITLE
Update Example autoencoder

### PR DESCRIPTION
**Pull request by ThomasDelteil**
## Description ##

- Remove symbolic implementation using dense layer and MNIST and created a gluon one using convolutional autoencoder and FashionMNIST

### CODE REVIEW

**Comment by roywei** [comment]
@ThomasDelteil Thanks for updating the example.
@mxnet-label-bot [Example, pr-awaiting-review]

File: example/autoencoder/convolutional_autoencoder.ipynb [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+   "source": [
+    "plt.figure(figsize=(20,4))\n",
+    "for i in range(10):\n",
+    "    idx = random.randint(0, len(test_dataset))\n",
+    "    img, _ = test_dataset[idx]\n",
+    "    x, _ = test_dataset_t[idx]\n",
+    "\n",
+    "    data = x.as_in_context(ctx).expand_dims(axis=0)\n",
+    "    output = net(data)\n",
+    "
```
Comment (thomelane) [inline example/autoencoder/convolutional_autoencoder.ipynb]: Wouldn't it be nicer to create these as separate HybridSequential blocks in the first place? And then AutoEncoder would be a HybridSequential block of encoder and decoder, and you can just reference encoder and decoder here. Also covers case when user changes the encoder and screws up the layer numbering.

Comment (thomelane) [inline example/autoencoder/convolutional_autoencoder.ipynb]: Kind of looks like we're interpolating along two axis here since 5x5 grid. Can we just have one line?
Maybe add some more images, have 10 smaller images in one line, like earlier on in notebook.

Comment (thomelane) [inline example/autoencoder/convolutional_autoencoder.ipynb]: taht -> taht

File: example/autoencoder/README.md [hunk 2]
```diff
@@ -1,16 +1,18 @@
...
+The idea of an autoencoder is to learn to use bottleneck architecture to encode the input and then try to decode it to reproduce the original. By doing so, the network learns to effectively compress the information of the input, the resulting embedding representation can then be used in several domains. For example as featurized representation for visual search, or in anomaly detection.
 
 ## Dataset
-The dataset used in this example is [MNIST](http://yann.lecun.com/exdb/mnist/) dataset. This example uses scikit-learn module to download this dataset.
 
-## Simple autoencoder example
-mnist_sae.py: this example uses a simple auto-encoder architecture to encode and decode MNIST images with size of 28x28 pixels. It contains several command line arguments. Pass -h (or --help) to view all available options. To start the training on CPU (use --gpu option for training on GPU) using default options:
+The dataset used in this example is [FashionMNIST](https://github.com/zalandoresearch/fashion-mnist) dataset. 
+
+## Variationnal Autoencoder
```
Comment (ankkhedia) [inline example/autoencoder/README.md]: Variational

File: example/autoencoder/convolutional_autoencoder.ipynb (hunk 1 above)
Comment (ThomasDelteil) [reply example/autoencoder/convolutional_autoencoder.ipynb]: agreed :+1:

**Comment by ankkhedia** [comment]
@aaronmarkham Another example to be reviewed :)

File: example/autoencoder/convolutional_autoencoder.ipynb (hunk 1 above)
Comment (ThomasDelteil) [reply example/autoencoder/convolutional_autoencoder.ipynb]: I think I disagree with that but I have updated as you suggested :+1:

**Comment by ThomasDelteil** [comment]
@ankkhedia @thomelane updated

**Comment by kalyc** [comment]
@ThomasDelteil there appears to be a CI build failure due to time out - could you please re-trigger the CI with an empty commit?

**Comment by kalyc** [comment]
@mxnet-label-bot update [pr-awaiting-merge]

File: example/autoencoder/convolutional_autoencoder.ipynb [hunk 3]
```diff
@@ -0,0 +1,587 @@
...
.../8YOXflf5d22N1zntrSuk7dJ+cO3duVqbp5PU84CnXtU+13/27S3HJ9MxSOr/oXPRxUBUXU39HRL4feJwZ3Vv1fv08rHumxzLpLiniPcaLniV8zOo80L/z36pl2gel5x8v02vo93o9fb7yeGKtvN5qu3pcXR3P3l66r6ntz1/a/jqfvZ7ORR8H+t26dvhZSdcO3zO1TG0/27c3zXGaAwAAAAAAAACADoWXQAAAAAAAAAAANaAp5WDuzq6ueuqS5dIZdd9yV2h1B1MXLXf5qkqx6q5mipepm5e66ZVkXfPmzcs+q4t3STLT2XKhe++9N9mTJk3KylRGc+CBB2Zl6srYv3//ZLt7aqNpnhVvA02pp5IOd6E86aSTKsu0D1VaeM0112T1xowZk2x3Da5K/e6/Q12IPUWrp33tDNRF2F2hlSqpT0TEb3/722T72K7C3bi171ya0yjqSnnPPfdU1lOXUJ17XlZy6+1sOZi2ua+FOp5VwhkR0atXr2SXZIoluZ9+n/aNz1n9O5VTuURH29WvUSULVKmff5ePR5UL6X14Sl5dO0oy0/ZEf4evh5/+9KeTve+++2ZlS5cuTXbJBXlN587rpSQz0rYsrYdOlfu9o+PJ+83vq7ujc1El2ttuu21WT9ukJLnX/vD0tHoNtTtKKtms+JlD57CWeb3bbrst2XfffXdWVnVe6GhK/f3ss88m288iffr0afN6JVlxM+HrTpVEy/cqPYP7eVyvWTVX2vr8etHruaRMpVx+FtD7rbL/G83av45LM/Vs6/2he5DuOb43VZ2R/GxSkojqNapCWUREbLfddsn+xS9+kZV1tGSoK9F2dTmYPnP585e2s47nkhysSpoXkfeNnmX8HkvXqPquiPy9hz4j+/NIez9b4AkEAAAAAAAAAFADeAkEAAAAAAAAAFADmlIO5u6FVVm63AVZ3d5dHqEukeo6565h7g5bRaMu63o9v6eSy6lmh9DrdbUru7a5t/+f/vSnNu2IvA9VjjJy5Mis3vDhw5M9bNiwymuU3GlVUqVZqqZOnZrVa7Svn3nmmWTvt99+WdmJJ56YbM9Ip3+nLroll+tHHnkkK1u4cGFDypaBqwAABytJREFU99ieNCpr0rHoLsgus2l21IXdJXkjRoxItraHZyTqbLdonQPeT9ofLgdasGBBskvyS51v7taq393oPNK/8fVT77
```
Comment (KellenSunderland) [inline example/autoencoder/convolutional_autoencoder.ipynb]: Nit: Would being more specific and saying 'encoder network' be helpful to readers?

File: example/autoencoder/convolutional_autoencoder.ipynb [hunk 4]
```diff
@@ -0,0 +1,587 @@
...
+    "from mxnet import autograd, nd, gluon"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Data\n",
+    "\n",
+    "We will use the FashionMNIST dataset which is of a similar format than MNIST but is richer and has more variance"
```
Comment (KellenSunderland) [inline example/autoencoder/convolutional_autoencoder.ipynb]: Supernit: "We will use the FashionMNIST dataset which is of a similar format than MNIST but is richer and has more variance" to me sounds more natural as "We will use the FashionMNIST dataset, which is of a similar format to MNIST but is richer and has more variance"

File: example/autoencoder/convolutional_autoencoder.ipynb (hunk 3 above)
Comment (KellenSunderland) [inline example/autoencoder/convolutional_autoencoder.ipynb]: Is net printed out here again just for the reader's reference?

File: example/autoencoder/convolutional_autoencoder.ipynb [hunk 5]
```diff
@@ -0,0 +1,587 @@
...
+    "            # Compute the L2 and L1 losses between the original and the generated image\n",
+    "            l2 = l2_loss(output.flatten(), data.flatten())\n",
+    "            l1 = l1_loss(output.flatten(), data.flatten())\n",
+    "            l =  l2 + l1 \n",
+    "        l.backward()\n",
+    "        trainer.step(data.shape[0])\n",
+    "        \n",
+    "        curr_loss += l.mean()\n",
+    "\n",
+    "    print(\"Epoch [{}], Loss {}\".format(e, curr_loss.asscalar()/(i+1)))"
```
Comment (KellenSunderland) [inline example/autoencoder/convolutional_autoencoder.ipynb]: Is the indenting off here?  Seems like i is out of scope.

File: example/autoencoder/convolutional_autoencoder.ipynb [hunk 6]
```diff
@@ -0,0 +1,587 @@
...
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import random\n",
+    "\n",
+    "import matplotlib.pyplot as plt\n",
+    "import mxnet as mx\n",
+    "from mxnet import autograd, nd, gluon"
```
Comment (KellenSunderland) [inline example/autoencoder/convolutional_autoencoder.ipynb]: Is nd used?

**Comment by stu1130** [comment]
@mxnet-label-bot update [pr-awaiting-response]

**Comment by roywei** [comment]
@ThomasDelteil Thanks for the contribution, could you address the comments and trigger CI?

**Comment by sandeep-krishnamurthy** [comment]
@ThomasDelteil - Can you please check CI failures and address review comments? Thanks

**Comment by anirudhacharya** [comment]
@ThomasDelteil ping for update :)

File: example/autoencoder/convolutional_autoencoder.ipynb (hunk 3 above)
Comment (ThomasDelteil) [reply example/autoencoder/convolutional_autoencoder.ipynb]: Not sure what you mean here, from the line you commented on, what we are plotting is not only the encoded representation but the encoding / decoding reconstruction?

File: example/autoencoder/convolutional_autoencoder.ipynb (hunk 4 above)
Comment (ThomasDelteil) [reply example/autoencoder/convolutional_autoencoder.ipynb]: Sure thing, thanks

File: example/autoencoder/convolutional_autoencoder.ipynb (hunk 3 above)
Comment (ThomasDelteil) [reply example/autoencoder/convolutional_autoencoder.ipynb]: yes removed it for clarity,m thanks

File: example/autoencoder/convolutional_autoencoder.ipynb (hunk 6 above)
Comment (ThomasDelteil) [reply example/autoencoder/convolutional_autoencoder.ipynb]: no, thanks, removed it

**Comment by ThomasDelteil** [comment]
@KellenSunderland is it good to merge now?
